data/README.md

```diff
@@ -0,0 +1,100 @@
+# Data
+
+A guide for managing a series of tubes.
+
+## Best Practices
+
+* Use stream processing for data which is not guaranteed to fit into memory.
+
+### Streaming
+
+* Avoid automatic retries for non-idempotent operations.
+* Avoid storing messages with different schemas in the same topic or message
+  queue.
+* Keep messages as small as possible.
+* Use as few distinct resources (CPU, Postgres, S3, HTTP requests) as possible
+  in each consumer.
+* Use at least once delivery with idempotent operations.
+* Use at most once delivery with non-idempotent operations.
+* Use constant memory to process streams when feasible.
+* When one message would perform more than on operation, instead have that
+  message append a new message for each operation.
+* When reading from the same queue with more than one consumer, ensure that it's
+  acceptable to process messages from that queue out of order.
+* When partitioning messages in a single topic, ensure that it's acceptable for
+  messages from different partitions to be processed out of order.
+
+## Glossary
+
+*Data Engineering*: the task of building infrastructure and process for
+ingesting, processing, and aggregating data so that it can be displayed to users
+or made available to data scientists.
+
+*Data Science*: the practice of using statistics, machine learning, and other
+tools to analyze data to discover trends and truths that can be used to provide
+business intelligence.
+
+*Batch Processing*: processing large amounts of data at once. This is acceptable
+for smaller amounts of data and can be simpler in terms of engineering and
+deployment. Some batch processes can also be useful for "recomputing the world"
+when you want to analyze existing data in a new way.
+
+*Data Streaming*: processing data in small chunks, one at a time, rather than
+processing all data at once. Streaming is necessary for processing infinite
+event streams. It's also useful for processing large amounts of data, because it
+prevents memory overflows during processing and makes it easier to process data
+in a distributed manner or real-time manner.
+
+*Real-time*: analyzing data and delivering results simultaneously so that stream
```

Comment:
This is "soft real time", as opposed to hard real time?

Reply:
Yes - "real time" is an almost worthless term because of how widely expectation for "real" varies, but it comes up often enough that it seemed useful to describe.